// @plures/praxis/dist/node/auth-STARLY7I.js
import {
  createSponsorsClient
} from "./chunk-DSDC2JWZ.js";
import {
  authenticateWithDeviceFlow
} from "./chunk-FXQZXAWF.js";
import "./chunk-QGM4M3NI.js";

// src/cli/commands/auth.ts
import * as fs from "fs";
import * as path from "path";
import * as os from "os";
var AUTH_DIR = path.join(os.homedir(), ".praxis");
var AUTH_FILE = path.join(AUTH_DIR, "auth.json");
var GITHUB_CLIENT_ID = "Ov23liQxF7P0BqUxVXHk";
function loadAuth() {
  try {
    if (fs.existsSync(AUTH_FILE)) {
      const data = fs.readFileSync(AUTH_FILE, "utf-8");
      return JSON.parse(data);
    }
  } catch (error) {
    console.warn("Failed to load authentication:", error);
  }
  return null;
}
function saveAuth(auth) {
  try {
    if (!fs.existsSync(AUTH_DIR)) {
      fs.mkdirSync(AUTH_DIR, { recursive: true });
    }
    fs.writeFileSync(AUTH_FILE, JSON.stringify(auth, null, 2), { mode: 384 });
    console.log(`
\u2713 Authentication saved to ${AUTH_FILE}`);
  } catch (error) {
    console.error("Failed to save authentication:", error);
    throw error;
  }
}
function deleteAuth() {
  try {
    if (fs.existsSync(AUTH_FILE)) {
      fs.unlinkSync(AUTH_FILE);
    }
  } catch (error) {
    console.warn("Failed to delete authentication:", error);
  }
}
async function loginCommand(options) {
  console.log("\n\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551   Praxis Cloud Authentication                     \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  const existingAuth = loadAuth();
  if (existingAuth) {
    console.log("\u26A0  Already authenticated");
    console.log(`   User: ${existingAuth.userLogin}`);
    console.log(
      `   Authenticated at: ${new Date(existingAuth.authenticatedAt).toLocaleString()}
`
    );
    console.log("Run 'praxis logout' to log out first.");
    return;
  }
  let authResult;
  if (options.token) {
    console.log("\u{1F510} Using provided personal access token...");
    try {
      const response = await fetch("https://api.github.com/user", {
        headers: {
          Authorization: `Bearer ${options.token}`,
          Accept: "application/vnd.github.v3+json"
        }
      });
      if (!response.ok) {
        throw new Error(`Invalid token: ${response.statusText}`);
      }
      const userData = await response.json();
      authResult = {
        success: true,
        token: options.token,
        user: {
          id: userData.id,
          login: userData.login,
          email: userData.email,
          name: userData.name,
          avatarUrl: userData.avatar_url
        }
      };
    } catch (error) {
      console.error("\n\u2717 Authentication failed");
      console.error(`  Error: ${error instanceof Error ? error.message : String(error)}`);
      process.exit(1);
    }
  } else {
    console.log("\u{1F510} Authenticating with GitHub device flow...");
    authResult = await authenticateWithDeviceFlow(GITHUB_CLIENT_ID);
  }
  if (!authResult.success || !authResult.token || !authResult.user) {
    console.error("\n\u2717 Authentication failed");
    process.exit(1);
  }
  console.log(`\u2713 Authenticated as ${authResult.user.login}`);
  console.log("\n\u{1F50D} Checking GitHub Sponsors status...");
  try {
    const sponsorsClient = createSponsorsClient(authResult.token);
    const subscription = await sponsorsClient.getSubscription(authResult.user.login);
    console.log(`\u2713 Subscription tier: ${subscription.tier}`);
    console.log(`  Status: ${subscription.status}`);
    console.log(`  Provider: ${subscription.provider}`);
    if (subscription.tier === "free") {
      console.log("\n\u{1F4A1} Upgrade to a paid tier for more features!");
      console.log("   Visit: https://github.com/sponsors/plures");
    }
  } catch (error) {
    console.warn("\n\u26A0  Could not check Sponsors status");
    console.warn(`   Error: ${error instanceof Error ? error.message : String(error)}`);
  }
  const storedAuth = {
    token: authResult.token,
    userId: authResult.user.id,
    userLogin: authResult.user.login,
    userName: authResult.user.name,
    userEmail: authResult.user.email,
    authenticatedAt: Date.now()
  };
  saveAuth(storedAuth);
  console.log("\n\u2713 Successfully logged in!");
  console.log("\nNext steps:");
  console.log("  \u2022 Use 'praxis cloud init' to set up cloud connection");
  console.log("  \u2022 Use 'praxis whoami' to check your authentication");
  console.log("  \u2022 Use 'praxis cloud status' to check subscription\n");
}
async function logoutCommand() {
  const auth = loadAuth();
  if (!auth) {
    console.log("\n\u2717 Not currently logged in");
    console.log("  Run 'praxis login' to authenticate\n");
    return;
  }
  console.log(`
Logging out user: ${auth.userLogin}...`);
  deleteAuth();
  console.log("\u2713 Successfully logged out\n");
}
async function whoamiCommand() {
  const auth = loadAuth();
  if (!auth) {
    console.log("\n\u2717 Not currently logged in");
    console.log("  Run 'praxis login' to authenticate\n");
    process.exit(1);
  }
  console.log("\n\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551   Current Authentication                          \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  console.log(`User: ${auth.userLogin}`);
  if (auth.userName) {
    console.log(`Name: ${auth.userName}`);
  }
  if (auth.userEmail) {
    console.log(`Email: ${auth.userEmail}`);
  }
  console.log(`User ID: ${auth.userId}`);
  console.log(`Authenticated: ${new Date(auth.authenticatedAt).toLocaleString()}`);
  console.log("\n\u{1F50D} Checking token validity...");
  try {
    const response = await fetch("https://api.github.com/user", {
      headers: {
        Authorization: `Bearer ${auth.token}`,
        Accept: "application/vnd.github.v3+json"
      }
    });
    if (response.ok) {
      console.log("\u2713 Token is valid");
      console.log("\n\u{1F50D} Checking subscription...");
      const sponsorsClient = createSponsorsClient(auth.token);
      const subscription = await sponsorsClient.getSubscription(auth.userLogin);
      console.log(`Tier: ${subscription.tier}`);
      console.log(`Status: ${subscription.status}`);
      console.log(`Provider: ${subscription.provider}`);
      console.log("\nLimits:");
      console.log(`  Syncs/month: ${subscription.limits.maxSyncsPerMonth.toLocaleString()}`);
      console.log(
        `  Storage: ${(subscription.limits.maxStorageBytes / 1024 / 1024).toFixed(0)} MB`
      );
      console.log(`  Apps: ${subscription.limits.maxApps}`);
      console.log(`  Team members: ${subscription.limits.maxTeamMembers ?? "Unlimited"}`);
      console.log(`  Support: ${subscription.limits.supportLevel}`);
    } else {
      console.log("\u2717 Token is invalid or expired");
      console.log("  Run 'praxis login' to re-authenticate");
    }
  } catch (error) {
    console.error("\n\u2717 Failed to check token validity");
    console.error(`  Error: ${error instanceof Error ? error.message : String(error)}`);
  }
  console.log();
}
function getAuthToken() {
  const auth = loadAuth();
  return auth?.token || null;
}
export {
  getAuthToken,
  loginCommand,
  logoutCommand,
  whoamiCommand
};
