// @plures/praxis/dist/node/build-Y7OT5VBF.js
import "./chunk-QGM4M3NI.js";

// src/cli/commands/build.ts
import { spawn } from "child_process";
import * as fs from "fs";
import * as path from "path";
async function build(options) {
  const output = options.output || "./dist";
  const target = options.target || "web";
  console.log("\n\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551   Praxis Production Build                         \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  const packageJsonPath = path.join(process.cwd(), "package.json");
  if (!fs.existsSync(packageJsonPath)) {
    console.error("Error: No package.json found in current directory.");
    console.log("Make sure you are in a Praxis project directory.\n");
    process.exit(1);
  }
  const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
  if (!packageJson.scripts?.build) {
    console.error('Error: No "build" script found in package.json.');
    console.log("\nTo build your Praxis project, add a build script to your package.json:");
    console.log('  "scripts": {');
    console.log('    "build": "vite build"');
    console.log("  }\n");
    process.exit(1);
  }
  console.log(`Building for: ${target}`);
  console.log(`Output: ${output}`);
  console.log("");
  if (target === "desktop") {
    await buildDesktop(options);
    return;
  }
  if (target === "mobile") {
    await buildMobile(options);
    return;
  }
  await buildWeb(options);
}
async function buildWeb(options) {
  const output = options.output || "./dist";
  console.log("\u{1F310} Building web application...\n");
  const env = {
    ...process.env,
    FORCE_COLOR: "1"
  };
  const args = ["run", "build"];
  if (output !== "./dist") {
    args.push("--", "--outDir", output);
  }
  return new Promise((resolve, reject) => {
    const child = spawn("npm", args, {
      stdio: "inherit",
      shell: true,
      cwd: process.cwd(),
      env
    });
    child.on("error", (error) => {
      console.error("Build failed:", error.message);
      reject(error);
    });
    child.on("exit", (code) => {
      if (code === 0) {
        console.log("\n\u2705 Build completed successfully!");
        console.log(`
Output: ${output}/`);
        console.log("\nTo preview the build:");
        console.log("  npm run preview\n");
        resolve();
      } else {
        console.error(`
\u274C Build failed with code ${code}`);
        process.exit(code || 1);
      }
    });
  });
}
async function buildDesktop(_options) {
  console.log("\u{1F5A5}\uFE0F  Building desktop application with Tauri...\n");
  const tauriConfigPath = path.join(process.cwd(), "src-tauri", "tauri.conf.json");
  if (!fs.existsSync(tauriConfigPath)) {
    console.log("Tauri is not configured for this project.");
    console.log("\nTo add Tauri support:");
    console.log("  1. npm install -D @tauri-apps/cli");
    console.log("  2. npm run tauri init");
    console.log("  3. praxis build --target desktop\n");
    console.log("For more information, see the Tauri integration docs.");
    process.exit(1);
  }
  const packageJson = JSON.parse(
    fs.readFileSync(path.join(process.cwd(), "package.json"), "utf-8")
  );
  const buildScript = packageJson.scripts?.["tauri:build"] || packageJson.scripts?.["tauri"];
  return new Promise((resolve, reject) => {
    const child = spawn("npm", ["run", buildScript ? "tauri:build" : "tauri", "build"], {
      stdio: "inherit",
      shell: true,
      cwd: process.cwd(),
      env: {
        ...process.env,
        FORCE_COLOR: "1"
      }
    });
    child.on("error", (error) => {
      console.error("Desktop build failed:", error.message);
      reject(error);
    });
    child.on("exit", (code) => {
      if (code === 0) {
        console.log("\n\u2705 Desktop build completed!");
        console.log("\nBundles are in: src-tauri/target/release/bundle/");
        resolve();
      } else {
        console.error(`
\u274C Desktop build failed with code ${code}`);
        process.exit(code || 1);
      }
    });
  });
}
async function buildMobile(_options) {
  console.log("\u{1F4F1} Building mobile application...\n");
  const tauriConfigPath = path.join(process.cwd(), "src-tauri", "tauri.conf.json");
  if (!fs.existsSync(tauriConfigPath)) {
    console.log("Tauri is not configured for this project.");
    console.log("\nTo add mobile support:");
    console.log("  1. npm install -D @tauri-apps/cli");
    console.log("  2. npm run tauri init");
    console.log("  3. npm run tauri android init");
    console.log("  4. npm run tauri ios init");
    console.log("  5. praxis build --target mobile\n");
    console.log("For more information, see the Tauri mobile docs.");
    process.exit(1);
  }
  console.log("Mobile build requires platform-specific setup:");
  console.log("\n\u{1F4F1} Android:");
  console.log("  npm run tauri android build");
  console.log("\n\u{1F34E} iOS:");
  console.log("  npm run tauri ios build\n");
}
export {
  build
};
