// @plures/praxis/dist/node/chunk-DSDC2JWZ.js
// src/cloud/billing.ts
var SubscriptionTier = /* @__PURE__ */ ((SubscriptionTier2) => {
  SubscriptionTier2["FREE"] = "free";
  SubscriptionTier2["SOLO"] = "solo";
  SubscriptionTier2["TEAM"] = "team";
  SubscriptionTier2["ENTERPRISE"] = "enterprise";
  return SubscriptionTier2;
})(SubscriptionTier || {});
var BillingProvider = /* @__PURE__ */ ((BillingProvider2) => {
  BillingProvider2["SPONSORS"] = "sponsors";
  BillingProvider2["MARKETPLACE"] = "marketplace";
  BillingProvider2["NONE"] = "none";
  return BillingProvider2;
})(BillingProvider || {});
var SubscriptionStatus = /* @__PURE__ */ ((SubscriptionStatus2) => {
  SubscriptionStatus2["ACTIVE"] = "active";
  SubscriptionStatus2["CANCELLED"] = "cancelled";
  SubscriptionStatus2["EXPIRED"] = "expired";
  SubscriptionStatus2["NONE"] = "none";
  return SubscriptionStatus2;
})(SubscriptionStatus || {});
var TIER_LIMITS = {
  ["free" /* FREE */]: {
    maxSyncsPerMonth: 1e3,
    maxStorageBytes: 10 * 1024 * 1024,
    // 10 MB
    maxTeamMembers: 1,
    maxApps: 1,
    supportLevel: "community"
  },
  ["solo" /* SOLO */]: {
    maxSyncsPerMonth: 5e4,
    maxStorageBytes: 1024 * 1024 * 1024,
    // 1 GB
    maxTeamMembers: 1,
    maxApps: 10,
    supportLevel: "standard"
  },
  ["team" /* TEAM */]: {
    maxSyncsPerMonth: 5e5,
    maxStorageBytes: 10 * 1024 * 1024 * 1024,
    // 10 GB
    maxTeamMembers: 10,
    maxApps: 50,
    supportLevel: "standard"
  },
  ["enterprise" /* ENTERPRISE */]: {
    maxSyncsPerMonth: 5e6,
    maxStorageBytes: 100 * 1024 * 1024 * 1024,
    // 100 GB
    maxTeamMembers: null,
    // unlimited
    maxApps: 1e3,
    supportLevel: "priority"
  }
};
function hasAccessToTier(subscription, requiredTier) {
  if (subscription.status !== "active" /* ACTIVE */) {
    return false;
  }
  const tierOrder = [
    "free" /* FREE */,
    "solo" /* SOLO */,
    "team" /* TEAM */,
    "enterprise" /* ENTERPRISE */
  ];
  const currentTierIndex = tierOrder.indexOf(subscription.tier);
  const requiredTierIndex = tierOrder.indexOf(requiredTier);
  return currentTierIndex >= requiredTierIndex;
}
function checkUsageLimits(subscription, usage) {
  const violations = [];
  if (usage.syncCount > subscription.limits.maxSyncsPerMonth) {
    violations.push(
      `Sync limit exceeded: ${usage.syncCount}/${subscription.limits.maxSyncsPerMonth}`
    );
  }
  if (usage.storageBytes > subscription.limits.maxStorageBytes) {
    violations.push(
      `Storage limit exceeded: ${(usage.storageBytes / 1024 / 1024).toFixed(2)}MB/${(subscription.limits.maxStorageBytes / 1024 / 1024).toFixed(2)}MB`
    );
  }
  if (subscription.limits.maxTeamMembers !== null && usage.teamMembers > subscription.limits.maxTeamMembers) {
    violations.push(
      `Team member limit exceeded: ${usage.teamMembers}/${subscription.limits.maxTeamMembers}`
    );
  }
  if (usage.appCount > subscription.limits.maxApps) {
    violations.push(`App limit exceeded: ${usage.appCount}/${subscription.limits.maxApps}`);
  }
  return {
    withinLimits: violations.length === 0,
    violations
  };
}
function createFreeSubscription() {
  return {
    tier: "free" /* FREE */,
    status: "active" /* ACTIVE */,
    provider: "none" /* NONE */,
    startDate: Date.now(),
    autoRenew: true,
    limits: TIER_LIMITS["free" /* FREE */]
  };
}
function createSponsorSubscription(tierName, monthlyPriceInCents) {
  let tier = "free" /* FREE */;
  if (monthlyPriceInCents >= 5e3) {
    tier = "enterprise" /* ENTERPRISE */;
  } else if (monthlyPriceInCents >= 2e3) {
    tier = "team" /* TEAM */;
  } else if (monthlyPriceInCents >= 500) {
    tier = "solo" /* SOLO */;
  }
  return {
    tier,
    status: "active" /* ACTIVE */,
    provider: "sponsors" /* SPONSORS */,
    sponsorTierId: tierName,
    startDate: Date.now(),
    autoRenew: true,
    limits: TIER_LIMITS[tier]
  };
}

// src/cloud/sponsors.ts
var GitHubSponsorsClient = class {
  token;
  accountLogin;
  constructor(token, accountLogin) {
    this.token = token;
    this.accountLogin = accountLogin;
  }
  /**
   * Get current user's sponsorship of the Praxis account
   */
  async getSponsorship(userLogin) {
    try {
      const query = `
        query($accountLogin: String!, $userLogin: String!) {
          user(login: $userLogin) {
            sponsorshipForViewerAsSponsor(activeOnly: true) {
              tier {
                id
                name
                monthlyPriceInCents
                description
                isOneTime
              }
              createdAt
              isActive
            }
            sponsorshipsAsSponsor(first: 100, activeOnly: true) {
              nodes {
                sponsorable {
                  ... on User {
                    login
                  }
                  ... on Organization {
                    login
                  }
                }
                tier {
                  id
                  name
                  monthlyPriceInCents
                  description
                  isOneTime
                }
                createdAt
                isActive
              }
            }
          }
        }
      `;
      const response = await fetch("https://api.github.com/graphql", {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.token}`,
          "Content-Type": "application/json"
        },
        body: JSON.stringify({
          query,
          variables: {
            accountLogin: this.accountLogin,
            userLogin
          }
        })
      });
      if (!response.ok) {
        throw new Error(`GitHub API error: ${response.statusText}`);
      }
      const data = await response.json();
      if (data.errors) {
        throw new Error(`GraphQL error: ${JSON.stringify(data.errors)}`);
      }
      const sponsorships = data.data?.user?.sponsorshipsAsSponsor?.nodes || [];
      const praxisSponsorship = sponsorships.find(
        (s) => s.sponsorable?.login === this.accountLogin
      );
      if (!praxisSponsorship) {
        return null;
      }
      return {
        sponsorLogin: userLogin,
        sponsorId: data.data.user.id,
        tier: {
          id: praxisSponsorship.tier.id,
          name: praxisSponsorship.tier.name,
          monthlyPriceInCents: praxisSponsorship.tier.monthlyPriceInCents,
          description: praxisSponsorship.tier.description,
          isOneTime: praxisSponsorship.tier.isOneTime
        },
        createdAt: praxisSponsorship.createdAt,
        isActive: praxisSponsorship.isActive
      };
    } catch (error) {
      console.error("Failed to get sponsorship:", error);
      return null;
    }
  }
  /**
   * Get subscription from sponsorship
   */
  async getSubscription(userLogin) {
    const sponsorship = await this.getSponsorship(userLogin);
    if (!sponsorship || !sponsorship.isActive) {
      return createFreeSubscription();
    }
    return createSponsorSubscription(sponsorship.tier.name, sponsorship.tier.monthlyPriceInCents);
  }
  /**
   * Check if user is a sponsor
   */
  async isSponsor(userLogin) {
    const sponsorship = await this.getSponsorship(userLogin);
    return sponsorship !== null && sponsorship.isActive;
  }
};
function createSponsorsClient(token, accountLogin = "plures") {
  return new GitHubSponsorsClient(token, accountLogin);
}

export {
  SubscriptionTier,
  BillingProvider,
  SubscriptionStatus,
  TIER_LIMITS,
  hasAccessToTier,
  checkUsageLimits,
  createFreeSubscription,
  createSponsorSubscription,
  GitHubSponsorsClient,
  createSponsorsClient
};
