// @plures/praxis/dist/node/cloud-AXOK4PSN.js
import {
  connectRelay
} from "./chunk-XCY2VIFX.js";
import {
  authenticateWithDeviceFlow
} from "./chunk-FXQZXAWF.js";
import "./chunk-QGM4M3NI.js";

// src/cli/commands/cloud.ts
import * as fs from "fs";
import * as path from "path";
var CONFIG_FILE = ".praxis-cloud.json";
var GITHUB_CLIENT_ID = "Ov23liQxF7P0BqUxVXHk";
function loadConfig() {
  try {
    const configPath = path.join(process.cwd(), CONFIG_FILE);
    if (fs.existsSync(configPath)) {
      const data = fs.readFileSync(configPath, "utf-8");
      return JSON.parse(data);
    }
  } catch (error) {
    console.warn("Failed to load cloud configuration:", error);
  }
  return null;
}
function saveConfig(config) {
  try {
    const configPath = path.join(process.cwd(), CONFIG_FILE);
    fs.writeFileSync(configPath, JSON.stringify(config, null, 2));
    console.log(`
\u2713 Configuration saved to ${CONFIG_FILE}`);
  } catch (error) {
    console.error("Failed to save cloud configuration:", error);
    throw error;
  }
}
async function cloudInit(options) {
  console.log("\n\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551   Welcome to Praxis Cloud Setup Wizard           \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  const existingConfig = loadConfig();
  if (existingConfig) {
    console.log("\u26A0  Existing cloud configuration found.");
    console.log(`   Endpoint: ${existingConfig.endpoint}`);
    console.log(`   App ID: ${existingConfig.appId}
`);
  }
  let endpoint = options.endpoint;
  if (!endpoint) {
    endpoint = "https://praxis-relay.azurewebsites.net";
    console.log(`Using default endpoint: ${endpoint}`);
  }
  let appId = options.appId;
  if (!appId) {
    appId = path.basename(process.cwd());
    console.log(`Using app ID from directory name: ${appId}`);
  }
  console.log("\n\u{1F510} Authenticating with GitHub...");
  const authResult = await authenticateWithDeviceFlow(GITHUB_CLIENT_ID);
  if (!authResult.success || !authResult.token) {
    console.error("\n\u2717 Authentication failed");
    process.exit(1);
  }
  console.log(`\u2713 Authenticated as ${authResult.user?.login || "unknown"}`);
  console.log("\n\u{1F517} Testing connection to Praxis Cloud...");
  try {
    const config = {
      endpoint,
      appId,
      authToken: authResult.token,
      autoSync: options.autoSync,
      syncInterval: options.interval ? parseInt(options.interval) : 5e3
    };
    const client = await connectRelay(endpoint, config);
    const health = await client.getHealth();
    if (health.status === "healthy") {
      console.log("\u2713 Connected successfully!");
      console.log(`  Status: ${health.status}`);
      console.log(`  Version: ${health.version}`);
    } else {
      console.log(`\u26A0  Connected but service is ${health.status}`);
    }
    await client.disconnect();
    const storedConfig = {
      endpoint,
      appId,
      authToken: authResult.token,
      autoSync: options.autoSync,
      syncInterval: config.syncInterval
    };
    saveConfig(storedConfig);
    console.log("\n\u2713 Praxis Cloud is now configured!");
    console.log("\nNext steps:");
    console.log("  \u2022 Use 'praxis cloud status' to check connection");
    console.log("  \u2022 Use 'praxis cloud sync' to manually sync");
    console.log("  \u2022 Use 'praxis cloud usage' to view metrics");
    console.log("\nIn your code:");
    console.log('  import { connectRelay } from "@plures/praxis/cloud";');
    console.log(`  const relay = await connectRelay("${endpoint}", {`);
    console.log(`    appId: "${appId}",`);
    console.log(`    authToken: "<your-token>"`);
    console.log("  });\n");
  } catch (error) {
    console.error("\n\u2717 Failed to connect to Praxis Cloud");
    console.error(`  Error: ${error instanceof Error ? error.message : String(error)}`);
    process.exit(1);
  }
}
async function cloudStatus() {
  const config = loadConfig();
  if (!config) {
    console.log("\n\u2717 No cloud configuration found");
    console.log("  Run 'praxis cloud init' to set up cloud connection\n");
    process.exit(1);
  }
  console.log("\n\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551   Praxis Cloud Status                             \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  console.log(`Endpoint: ${config.endpoint}`);
  console.log(`App ID: ${config.appId}`);
  console.log(`Auto Sync: ${config.autoSync ? "enabled" : "disabled"}`);
  if (config.autoSync) {
    console.log(`Sync Interval: ${config.syncInterval}ms`);
  }
  try {
    const client = await connectRelay(config.endpoint, config);
    const health = await client.getHealth();
    console.log(`
Connection: \u2713 Connected`);
    console.log(`Status: ${health.status}`);
    console.log(`Version: ${health.version}`);
    console.log("\nServices:");
    console.log(`  Relay: ${health.services.relay ? "\u2713" : "\u2717"}`);
    console.log(`  Event Grid: ${health.services.eventGrid ? "\u2713" : "\u2717"}`);
    console.log(`  Storage: ${health.services.storage ? "\u2713" : "\u2717"}`);
    console.log(`  Auth: ${health.services.auth ? "\u2713" : "\u2717"}`);
    await client.disconnect();
  } catch (error) {
    console.log(`
Connection: \u2717 Failed`);
    console.log(`Error: ${error instanceof Error ? error.message : String(error)}`);
  }
  console.log();
}
async function cloudSync() {
  const config = loadConfig();
  if (!config) {
    console.log("\n\u2717 No cloud configuration found");
    console.log("  Run 'praxis cloud init' to set up cloud connection\n");
    process.exit(1);
  }
  console.log("\n\u{1F504} Syncing to Praxis Cloud...");
  try {
    const client = await connectRelay(config.endpoint, config);
    await client.sync({
      type: "delta",
      appId: config.appId,
      clock: {},
      facts: [],
      events: [],
      timestamp: Date.now()
    });
    console.log("\u2713 Sync completed successfully\n");
    await client.disconnect();
  } catch (error) {
    console.error("\n\u2717 Sync failed");
    console.error(`  Error: ${error instanceof Error ? error.message : String(error)}
`);
    process.exit(1);
  }
}
async function cloudUsage() {
  const config = loadConfig();
  if (!config) {
    console.log("\n\u2717 No cloud configuration found");
    console.log("  Run 'praxis cloud init' to set up cloud connection\n");
    process.exit(1);
  }
  console.log("\n\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551   Praxis Cloud Usage Metrics                      \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  try {
    const client = await connectRelay(config.endpoint, config);
    const usage = await client.getUsage();
    console.log(`App ID: ${usage.appId}`);
    console.log(`
Metrics:`);
    console.log(`  Total Syncs: ${usage.syncCount}`);
    console.log(`  Events Forwarded: ${usage.eventCount}`);
    console.log(`  Facts Synced: ${usage.factCount}`);
    console.log(`  Storage Used: ${(usage.storageBytes / 1024).toFixed(2)} KB`);
    const periodDuration = usage.periodEnd - usage.periodStart;
    const durationHours = (periodDuration / 1e3 / 60 / 60).toFixed(1);
    console.log(`
Period: ${durationHours} hours`);
    console.log(`  From: ${new Date(usage.periodStart).toLocaleString()}`);
    console.log(`  To: ${new Date(usage.periodEnd).toLocaleString()}`);
    console.log();
    await client.disconnect();
  } catch (error) {
    console.error("\n\u2717 Failed to retrieve usage metrics");
    console.error(`  Error: ${error instanceof Error ? error.message : String(error)}
`);
    process.exit(1);
  }
}
export {
  cloudInit,
  cloudStatus,
  cloudSync,
  cloudUsage
};
