// @plures/praxis/dist/node/create-TRLSVCNQ.js
import "./chunk-QGM4M3NI.js";

// src/cli/commands/create.ts
import { mkdir, writeFile } from "fs/promises";
import { join, resolve } from "path";
import { existsSync } from "fs";
function generatePackageJson(name) {
  const pkg = {
    name,
    version: "0.1.0",
    private: true,
    type: "module",
    scripts: {
      dev: "vite",
      build: "tsc && vite build",
      preview: "vite preview",
      generate: "praxis generate",
      test: "vitest"
    },
    dependencies: {
      "@plures/praxis": "^0.2.1"
    },
    devDependencies: {
      "@sveltejs/vite-plugin-svelte": "^5.0.0",
      "@tsconfig/svelte": "^5.0.4",
      svelte: "^5.0.0",
      "svelte-check": "^4.0.0",
      typescript: "^5.6.0",
      vite: "^6.0.0",
      vitest: "^4.0.0"
    }
  };
  return JSON.stringify(pkg, null, 2);
}
function generateTsConfig() {
  const config = {
    extends: "@tsconfig/svelte/tsconfig.json",
    compilerOptions: {
      target: "ESNext",
      module: "ESNext",
      moduleResolution: "bundler",
      strict: true,
      esModuleInterop: true,
      skipLibCheck: true,
      resolveJsonModule: true,
      isolatedModules: true,
      noEmit: true
    },
    include: ["src/**/*"],
    exclude: ["node_modules"]
  };
  return JSON.stringify(config, null, 2);
}
function generateViteConfig() {
  return `import { defineConfig } from 'vite';
import { svelte } from '@sveltejs/vite-plugin-svelte';

export default defineConfig({
  plugins: [svelte()],
  resolve: {
    alias: {
      '@': '/src'
    }
  }
});
`;
}
function generateMainTs(name) {
  return `/**
 * ${name} - Built with Praxis
 */
import App from './App.svelte';
import { mount } from 'svelte';

const app = mount(App, {
  target: document.getElementById('app')!
});

export default app;
`;
}
function generateAppSvelte(name) {
  return `<script lang="ts">
  import { createPraxisEngine, PraxisRegistry, defineFact, defineEvent, defineRule } from '@plures/praxis';

  // Define your context type
  interface AppContext {
    initialized: boolean;
    message: string;
  }

  // Create the registry and engine
  const registry = new PraxisRegistry<AppContext>();
  
  const engine = createPraxisEngine({
    initialContext: {
      initialized: true,
      message: 'Welcome to ${name}!'
    },
    registry
  });

  // Reactive context
  let context = $state(engine.getContext());
</script>

<main>
  <h1>{context.message}</h1>
  <p>Your Praxis application is running!</p>
  
  <div class="card">
    <h2>Next Steps</h2>
    <ul>
      <li>Define your schema in <code>src/schemas/app.schema.js</code></li>
      <li>Run <code>praxis generate</code> to generate code</li>
      <li>Build your application logic with facts, events, and rules</li>
    </ul>
  </div>
</main>

<style>
  main {
    max-width: 800px;
    margin: 0 auto;
    padding: 2rem;
    font-family: system-ui, -apple-system, sans-serif;
  }

  h1 {
    color: #333;
    text-align: center;
  }

  .card {
    background: #f5f5f5;
    border-radius: 8px;
    padding: 1.5rem;
    margin-top: 2rem;
  }

  .card h2 {
    margin-top: 0;
  }

  code {
    background: #e0e0e0;
    padding: 0.2rem 0.4rem;
    border-radius: 4px;
    font-size: 0.9em;
  }

  li {
    margin: 0.5rem 0;
  }
</style>
`;
}
function generateIndexHtml(name) {
  return `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>${name}</title>
    <style>
      * {
        margin: 0;
        padding: 0;
        box-sizing: border-box;
      }
      body {
        min-height: 100vh;
        background: linear-gradient(135deg, #f5f7fa 0%, #c3cfe2 100%);
      }
    </style>
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="/src/main.ts"></script>
  </body>
</html>
`;
}
function generateAppSchema(name) {
  return `/**
 * ${name} Schema
 * 
 * Define your application schema here.
 * Run 'praxis generate' to generate code from this schema.
 */

export const schema = {
  version: '1.0.0',
  name: '${name}',
  description: 'A Praxis application',
  
  models: [
    {
      name: 'Item',
      description: 'Example item model',
      fields: [
        { name: 'id', type: 'string' },
        { name: 'name', type: 'string' },
        { name: 'createdAt', type: 'date' }
      ]
    }
  ],
  
  components: [
    {
      name: 'ItemForm',
      type: 'form',
      description: 'Form for creating items',
      model: 'Item'
    },
    {
      name: 'ItemList',
      type: 'list',
      description: 'List of items',
      model: 'Item'
    }
  ],
  
  logic: [
    {
      id: 'item-logic',
      description: 'Item management logic',
      
      events: [
        {
          tag: 'CREATE_ITEM',
          payload: { name: 'string' },
          description: 'Create a new item'
        },
        {
          tag: 'DELETE_ITEM',
          payload: { itemId: 'string' },
          description: 'Delete an item'
        }
      ],
      
      facts: [
        {
          tag: 'ItemCreated',
          payload: { itemId: 'string', name: 'string' },
          description: 'An item was created'
        },
        {
          tag: 'ItemDeleted',
          payload: { itemId: 'string' },
          description: 'An item was deleted'
        }
      ],
      
      rules: [
        {
          id: 'item.create',
          description: 'Create a new item',
          priority: 10
        }
      ],
      
      constraints: [
        {
          id: 'item.name-required',
          description: 'Item name must not be empty',
          message: 'Item name is required'
        }
      ]
    }
  ]
};

export default schema;
`;
}
function generateSvelteConfig() {
  return `import { vitePreprocess } from '@sveltejs/vite-plugin-svelte';

export default {
  preprocess: vitePreprocess()
};
`;
}
function generateReadme(name, template) {
  return `# ${name}

A Praxis application created with the ${template} template.

## Getting Started

### Install dependencies

\`\`\`bash
npm install
\`\`\`

### Start development server

\`\`\`bash
npm run dev
\`\`\`

### Generate code from schema

\`\`\`bash
npm run generate
\`\`\`

### Build for production

\`\`\`bash
npm run build
\`\`\`

## Project Structure

\`\`\`
${name}/
\u251C\u2500\u2500 src/
\u2502   \u251C\u2500\u2500 schemas/
\u2502   \u2502   \u2514\u2500\u2500 app.schema.js    # Application schema
\u2502   \u251C\u2500\u2500 components/          # Generated & custom components
\u2502   \u251C\u2500\u2500 logic/               # Logic engine setup
\u2502   \u251C\u2500\u2500 App.svelte           # Root component
\u2502   \u2514\u2500\u2500 main.ts              # Application entry
\u251C\u2500\u2500 public/                  # Static assets
\u251C\u2500\u2500 package.json
\u251C\u2500\u2500 tsconfig.json
\u251C\u2500\u2500 vite.config.ts
\u2514\u2500\u2500 README.md
\`\`\`

## Documentation

- [Praxis Framework Guide](https://github.com/plures/praxis/blob/main/FRAMEWORK.md)
- [Getting Started](https://github.com/plures/praxis/blob/main/GETTING_STARTED.md)
- [Schema Reference](https://github.com/plures/praxis/blob/main/docs/guides/schema.md)

## License

MIT
`;
}
function generateGitignore() {
  return `# Dependencies
node_modules/

# Build output
dist/
build/
.svelte-kit/

# Generated code
generated/

# Environment
.env
.env.local
.env.*.local

# IDE
.vscode/
.idea/
*.swp
*.swo

# OS
.DS_Store
Thumbs.db

# Logs
*.log
npm-debug.log*

# Test coverage
coverage/
`;
}
async function createApp(name, options) {
  const template = options.template || "basic";
  const appDir = resolve(process.cwd(), name);
  if (existsSync(appDir)) {
    console.error(`
\u2717 Directory '${name}' already exists`);
    console.error("  Please choose a different name or remove the existing directory.\n");
    process.exit(1);
  }
  console.log(`
\u{1F4E6} Creating Praxis application: ${name}`);
  console.log(`   Template: ${template}`);
  console.log(`   Location: ${appDir}
`);
  const dirs = [
    appDir,
    join(appDir, "src"),
    join(appDir, "src", "schemas"),
    join(appDir, "src", "components"),
    join(appDir, "src", "logic"),
    join(appDir, "src", "store"),
    join(appDir, "public")
  ];
  for (const dir of dirs) {
    await mkdir(dir, { recursive: true });
  }
  console.log("\u2713 Created directory structure");
  await writeFile(join(appDir, "package.json"), generatePackageJson(name));
  await writeFile(join(appDir, "tsconfig.json"), generateTsConfig());
  await writeFile(join(appDir, "vite.config.ts"), generateViteConfig());
  await writeFile(join(appDir, "svelte.config.js"), generateSvelteConfig());
  await writeFile(join(appDir, ".gitignore"), generateGitignore());
  await writeFile(join(appDir, "README.md"), generateReadme(name, template));
  console.log("\u2713 Created configuration files");
  await writeFile(join(appDir, "index.html"), generateIndexHtml(name));
  await writeFile(join(appDir, "src", "main.ts"), generateMainTs(name));
  await writeFile(join(appDir, "src", "App.svelte"), generateAppSvelte(name));
  await writeFile(join(appDir, "src", "schemas", "app.schema.js"), generateAppSchema(name));
  console.log("\u2713 Created source files");
  await writeFile(
    join(appDir, "src", "components", ".gitkeep"),
    "# Generated components will be placed here\n"
  );
  await writeFile(join(appDir, "src", "logic", ".gitkeep"), "# Logic files will be placed here\n");
  await writeFile(join(appDir, "src", "store", ".gitkeep"), "# Store files will be placed here\n");
  await writeFile(join(appDir, "public", ".gitkeep"), "# Static assets go here\n");
  console.log("\u2713 Created placeholder files");
  console.log(`
\u2705 Application created successfully!

Next steps:

  cd ${name}
  npm install
  npm run dev

Then open http://localhost:5173 in your browser.

To generate code from your schema:

  npm run generate

For more information, see the README.md in your project.
`);
}
function generateComponentSchema(name) {
  return `/**
 * ${name} Component Schema
 */

export const componentSchema = {
  name: '${name}',
  type: 'display',
  description: 'A ${name} component',
  
  props: [
    { name: 'title', type: 'string', required: true },
    { name: 'content', type: 'string', required: false }
  ],
  
  events: [
    { name: 'click', description: 'Emitted when clicked' }
  ]
};

export default componentSchema;
`;
}
function generateComponentSvelte(name) {
  return `<script lang="ts">
  /**
   * ${name} Component
   * 
   * Generated by Praxis.
   */
  
  interface Props {
    title: string;
    content?: string;
  }
  
  let { title, content = '' }: Props = $props();
</script>

<div class="${name.toLowerCase()}" role="button" tabindex="0">
  <h3>{title}</h3>
  {#if content}
    <p>{content}</p>
  {/if}
  <slot />
</div>

<style>
  .${name.toLowerCase()} {
    padding: 1rem;
    border: 1px solid #ddd;
    border-radius: 8px;
    margin: 0.5rem 0;
    cursor: pointer;
    transition: box-shadow 0.2s;
  }

  .${name.toLowerCase()}:hover {
    box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  }

  h3 {
    margin: 0 0 0.5rem 0;
    color: #333;
  }

  p {
    margin: 0;
    color: #666;
  }
</style>
`;
}
async function createComponent(name, options) {
  const outputDir = options.directory || join(process.cwd(), "src", "components");
  const componentDir = join(outputDir, name);
  if (existsSync(componentDir)) {
    console.error(`
\u2717 Component '${name}' already exists at ${componentDir}`);
    console.error("  Please choose a different name or remove the existing component.\n");
    process.exit(1);
  }
  console.log(`
\u{1F9E9} Creating component: ${name}`);
  console.log(`   Location: ${componentDir}
`);
  await mkdir(componentDir, { recursive: true });
  await writeFile(join(componentDir, `${name}.svelte`), generateComponentSvelte(name));
  await writeFile(join(componentDir, `${name}.schema.js`), generateComponentSchema(name));
  await writeFile(
    join(componentDir, "index.ts"),
    `export { default as ${name} } from './${name}.svelte';
`
  );
  console.log(`\u2713 Created ${name}.svelte`);
  console.log(`\u2713 Created ${name}.schema.js`);
  console.log(`\u2713 Created index.ts`);
  console.log(`
\u2705 Component created successfully!

Import it in your application:

  import { ${name} } from './components/${name}';

Or use the schema for generation:

  praxis generate --schema src/components/${name}/${name}.schema.js
`);
}
async function create(type, name, options) {
  if (!name) {
    console.error("\n\u2717 Please provide a name for the " + type);
    console.error(`  Usage: praxis create ${type} <name>
`);
    process.exit(1);
  }
  if (!/^[a-zA-Z][a-zA-Z0-9-_]*$/.test(name)) {
    console.error("\n\u2717 Invalid name format");
    console.error(
      "  Name must start with a letter and contain only letters, numbers, hyphens, and underscores.\n"
    );
    process.exit(1);
  }
  switch (type) {
    case "app":
      await createApp(name, options);
      break;
    case "component":
      await createComponent(name, options);
      break;
    default:
      console.error(`
\u2717 Unknown type: ${type}`);
      console.error("  Valid types: app, component\n");
      process.exit(1);
  }
}
export {
  create
};
