// @plures/praxis/dist/node/dev-PMJZUYGE.js
import "./chunk-QGM4M3NI.js";

// src/cli/commands/dev.ts
import { spawn } from "child_process";
import * as fs from "fs";
import * as path from "path";
async function dev(options) {
  const port = options.port || "5173";
  const host = options.host || "localhost";
  console.log("\n\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551   Praxis Development Server                       \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  const packageJsonPath = path.join(process.cwd(), "package.json");
  const viteConfigPath = path.join(process.cwd(), "vite.config.ts");
  void viteConfigPath;
  if (!fs.existsSync(packageJsonPath)) {
    console.error("Error: No package.json found in current directory.");
    console.log("Make sure you are in a Praxis project directory.\n");
    console.log("To create a new project, run:");
    console.log("  praxis create app my-app\n");
    process.exit(1);
  }
  const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
  if (!packageJson.scripts?.dev) {
    console.error('Error: No "dev" script found in package.json.');
    console.log("\nTo use the Praxis dev server, add a dev script to your package.json:");
    console.log('  "scripts": {');
    console.log('    "dev": "vite"');
    console.log("  }\n");
    process.exit(1);
  }
  console.log(`Starting development server on http://${host}:${port}
`);
  const args = ["run", "dev", "--", "--port", port, "--host", host];
  if (options.open) {
    args.push("--open");
  }
  const child = spawn("npm", args, {
    stdio: "inherit",
    shell: true,
    cwd: process.cwd(),
    env: {
      ...process.env,
      FORCE_COLOR: "1"
    }
  });
  child.on("error", (error) => {
    console.error("Failed to start development server:", error.message);
    process.exit(1);
  });
  child.on("exit", (code) => {
    if (code !== 0 && code !== null) {
      console.error(`Development server exited with code ${code}`);
      process.exit(code);
    }
  });
  process.on("SIGINT", () => {
    console.log("\n\nStopping development server...");
    child.kill("SIGTERM");
    process.exit(0);
  });
}
export {
  dev
};
